### PR TITLE
Add `get_cosine_scaled_reward`

### PR DESCRIPTION
**Pull request by qgallouedec**
TRL ships only a handful of built-in reward functions. The RLVR/GRPO literature describes many more rule-based, verifiable rewards that users currently reimplement themselves. This PR continues closing that gap (after the n-gram repetition penalty) with a length-aware reward for concise reasoning.

## What

Adds `get_cosine_scaled_reward`, which scales a correctness reward by completion length following a cosine schedule:

$$R_{\text{cosine}}(y) = v_{\min} + \tfrac{1}{2}(v_{\max} - v_{\min})\left(1 + \cos\left(\tfrac{|y|}{L_{\max}}\pi\right)\right)$$

- A **correct** completion is rewarded more when it is **shorter**.
- A **wrong** completion is penalized less when it is **longer** (preserving exploration).
- An unparseable gold solution is skipped (reward `None`), as in `accuracy_reward`.

```python
from trl.rewards import get_cosine_scaled_reward

cosine_scaled_reward = get_cosine_scaled_reward(max_len=4096)  # max_len in tokens
trainer = GRPOTrainer(..., reward_funcs=[cosine_scaled_reward])
```

This is distinct from the existing `get_soft_overlong_punishment` (DAPO), which is a flat over-budget cap; this one shapes the correctness reward across the whole length range. It pairs naturally with a repetition penalty, since length shaping can otherwise be reward-hacked with degenerate repetition.

Reference: Appendix C.1 of [*Demystifying Long Chain-of-Thought Reasoning in LLMs*](https://huggingface.co/papers/2502.03373).

## Notes

- Correctness reuses `accuracy_reward` (math verification), so it stays consistent with TRL's existing accuracy semantics rather than duplicating the parsing/verification logic.
- Length is measured in **tokens** via `completion_ids` (consistent with `get_soft_overlong_punishment`), not characters.
- Implemented as a picklable callable class, matching `get_soft_overlong_punishment` (required by the async GRPO rollout worker).
- `max_len` is required (no default), since a sensible value depends on the token budget; the value bounds keep the paper/Open-R1 defaults.
- Adds a dedicated subsection to `paper_index.md` and an entry to the rewards docs.
- 8 unit tests covering correct/wrong length monotonicity, exact midpoint and boundary values, custom bounds, unparseable-gold skipping, and picklability.

---

> [!NOTE]
> **Low Risk**
> Additive reward API and documentation with unit tests; no changes to trainer core paths beyond a new optional reward function.
> 
> **Overview**
> Adds **`get_cosine_scaled_reward`**, a built-in GRPO/RLOO reward that combines **`accuracy_reward`** math verification with **token-length shaping** via a cosine schedule (paper Appendix C.1): shorter **correct** completions score higher; longer **wrong** ones are penalized less. Length uses **`completion_ids`**; lengths above **`max_len`** are clamped so the schedule does not wrap. Implemented as picklable **`_CosineScaledReward`** (same pattern as other factory rewards).
> 
> **Exports** from `trl.rewards` and **autodoc** in `rewards.md`. **`paper_index.md`** now documents both cosine length scaling and repetition penalty as a paired recipe. **Eight tests** cover monotonicity, boundaries, clamping, unparseable gold (`None`), custom bounds, and pickling.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 489d7ee091374f80ce9b35e87a25cca41defc6f3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/rewards/accuracy_rewards.py [hunk 1]
```diff
@@ -114,6 +115,109 @@ def accuracy_reward(
...
+                rewards.append(None)
+                continue
+            progress = len(ids) / self.max_len
+            cosine = math.cos(progress * math.pi)
+            if correct:
+                min_value, max_value = self.min_value_correct, self.max_value_correct
+            else:
+                # Swap the bounds so that a shorter wrong completion is penalized more than a longer one.
+                min_value, max_value = self.max_value_wrong, self.min_value_wrong
+            rewards.append(float(min_value + 0.5 * (max_value - min_value) * (1.0 + cosine)))
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/rewards/accuracy_rewards.py:219]: I think this issue needs to be addressed: cosine wraps for over-length completions.

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks.

I think the reported bug by Cursor needs to be addressed. Additionally, there are conflicts with the main branch that should be fixed.

File: trl/rewards/accuracy_rewards.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/rewards/accuracy_rewards.py:219]: solved

**Review by albertvillanova — APPROVED** [review]
